### ISSUE TITLE
question: serve stale records via cache when backend is down

### ISSUE THREAD

**Issue by JarenGlover** [coredns/coredns#3158]
At a high level, I am trying to understand if coredns can serve stale data when the backend/upstreams fails.

An example of this can be find via [Unbound](https://nlnetlabs.nl/projects/unbound/about/)'s [serve-expired](http://manpages.ubuntu.com/manpages/bionic/man5/unbound.conf.5.html) configuration option.

![image](https://user-images.githubusercontent.com/1428354/63303259-28627400-c294-11e9-9283-0eef70055361.png)

In short, I want to be able to serve stale results via the cache when the backend fail during an outage.

More [context](https://github.com/vttale/serve-stale) on why one might want to serve stale data.

**Comment by uruddarraju** [comment]
+1. I think this is an important feature to have if someone wants to run an entire dns infrastructure just on coredns, like we are trying to do.

**Comment by JarenGlover** [comment]
It appears that we are checking if TTL is greater than 0 [here](https://github.com/coredns/coredns/blob/master/plugin/cache/cache.go#L173). If true, then it will never return from cache when TTL is 0.

**Comment by miekg** [comment]
currently cache doesn't implement this feature. Would be nice though; like how unbound implemented it. Doesn't seem that hard actually

**Comment by chrisohaver** [comment]
> like how unbound implemented it. Doesn't seem that hard actually.

Ah, yes, as explained in the bind description. Inferring from that, it does change behavior slightly; It will serve TTL 0 records occasionally when the backend is _not_ down.  I think this is a fairly minor behavior change, but probably should make the feature optional for that reason.  

Thinking on it: It would require that _cache_ continue to pass the query down the plugin chain when sending a TTL 0 record to a client, so the backend plugin can process it and write to cache if the backend is back up.  When the backend plugin writes its answer, we would need to know if an answer was already written to the client, and if so just write to cache without sending a second answer to the client.

**Comment by soffchen** [comment]
> At a high level, I am trying to understand if coredns can serve stale data when the backend/upstreams fails.
> 
> An example of this can be find via [Unbound](https://nlnetlabs.nl/projects/unbound/about/)'s [serve-expired](http://manpages.ubuntu.com/manpages/bionic/man5/unbound.conf.5.html) configuration option.
> 
> ![image](https://user-images.githubusercontent.com/1428354/63303259-28627400-c294-11e9-9283-0eef70055361.png)
> 
> In short, I want to be able to serve stale results via the cache when the backend fail during an outage.
> 
> More [context](https://github.com/vttale/serve-stale) on why one might want to serve stale data.

Not only when upstream downs.

This feature of unbound will return zero ttl record from cache first with 0ms response to client, and refresh the record from upstream asynchronously.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] question: ser..." ]
>> like how unbound implemented it. Doesn't seem that hard actually.
>
>Ah, yes, as explained in the bind description. Inferring from that, it does change behavior slightly; It will serve TTL 0 records occasionally when the backend is _not_ down.  I think this is a fairly minor behavior change, but probably should make the feature optional for that reason.

didn't we always round up to 5s? Or was that dropped at some point. Hazy memory 
of something happening in this space a while back

>
>Thinking on it: It would require that _cache_ continue to pass the query down the plugin chain when sending a TTL 0 record to a client, so the backend plugin can process it and write to cache if the backend is back up.  When the backend plugin writes its answer, we would need to know if an answer was already written to the client, and if so just write to cache without sending a second answer to the client.

yep, althought may be difficult(-ish) to pull off without creating a race in the 
response writer.

/Miek

--
Miek Gieben

**Comment by JarenGlover** [comment]
@miekg @chrisohaver thanks for the quick response. I will try to find the bind (and or unbound) reference material to see if that helps with inspiration. 

However, how would you today suggest someone to tolerate a backend failure? (i.e. server stale results while the backend is down)

We could increase the cached TTL and increase the prefetch ... However, a aggressive prefetch might increase the load on the backend (in a world where you have multiple coredns servers (resolvers) but a single central backend)   

Another option I am going to read up more on is using the redisc plugin but not sure that will give me the experience I want. 

Thanks in advance.

**Comment by chrisohaver** [comment]
> didn't we always round up to 5s? Or was that dropped at some point. Hazy memory
of something happening in this space a while back

Yes, we default to a minimum TTL of 5. Meaning that when we write the entry to the cache, we round up to 5.  But we still serve records from cache with the TTL adjusted from the time it was stored - e.g. 3 seconds after writing an entry to cache with TTL 5, the TTL would be 2.

So if we implement it in the way the feature is described in the unbound doc, it will would always serve 0 TTL cache entry to the client, regardless of the status of the backend.  IOW once the backend comes back up,  we will continue to serve the stale record until the cache entry is updated.

**Comment by stp-ip** [comment]
Fixed by #3468 
/close
